### PR TITLE
fix(project-host-setup): allow adding a local host to a runtime-only project

### PR DESCRIPTION
**Pull request by PannenetsF**
## Summary

Fixes a failure when adding a **local host** to a project that exists **only on a remote Orca runtime**.

**Problem** — With a remote Orca server hosting a project such as `github:pytorch/pytorch`, adding a local host to that project via *Available Hosts* failed with:

```
Error invoking remote method 'projectHostSetups:setupExistingFolder': Error: Project not found: github:pytorch/pytorch
```

**Root cause** — The local IPC handler `projectHostSetups:setupExistingFolder` (`src/main/ipc/repos.ts`) had an upfront guard requiring the project to already exist in the local desktop store (`store.getProjects()`). A project created on the remote runtime has no local project record until a folder is imported, so the guard rejected the legitimate operation. The symmetric runtime handler (`orca-runtime.ts` → `setupProjectExistingFolder`) never had this guard, making the two inconsistent.

**Fix**
- `src/shared/project-host-setup-projection.ts` — add `getProjectIdForProviderIdentity`, deriving the project id from a `ProjectProviderIdentity` using the same normalization as `getProjectIdentityKey`, so identity can be reconstructed without a local project record.
- `src/main/ipc/repos.ts` — remove the premature "Project not found" guard; `alignRepoWithRequestedProject` now falls back to a `projectProviderIdentity` carried on the request when there is no local project. The import is made **atomic**: if alignment fails, a freshly-imported repo is rolled back (`store.removeProject` + cache invalidation) so a mismatched import no longer orphans a repo.
- `src/main/runtime/orca-runtime.ts` — apply the same identity-from-request fallback (symmetric direction). Refactored into `completeProjectHostSetup` / `linkRepoToProjectHostSetup` so both `setupProjectExistingFolder` and `setupProjectClone` share the rollback-on-failure path.
- `src/main/runtime/rpc/methods/project-runtime-rpc-methods.ts` + `src/shared/types.ts` — thread the optional `projectProviderIdentity` (github provider only) through the RPC schemas and shared arg types.
- `src/renderer/src/store/slices/repos.ts` — when the target host may lack a project record, attach the selected source-host identity (from args or the local `projects` list) before crossing the local/runtime boundary.

## Screenshots

No visual change. All edits are in main-process IPC handlers, the runtime service, RPC schemas, shared projection/types, and store-slice routing logic. No renderer UI/layout is modified.

## Testing

- [ ] `pnpm lint`
- [ ] `pnpm typecheck`
- [ ] `pnpm test`
- [ ] `pnpm build`
- [x] Added or updated high-quality tests for my changes, or explained why they aren't needed

The full `pnpm` checklist scripts were not run in the authoring environment; targeted checks were run instead and pass: `tsc` (node + web) and `oxlint` on the changed files are clean, and the affected suites (`project-host-setup-projection`, `repos-remote`, `orca-runtime`, `persistence` — 1260 tests) pass. `pnpm build` was not run. Please run the full checklist in CI.

New tests: unit tests for `getProjectIdForProviderIdentity` (normalization/host variants); IPC handler tests (setup for a runtime-only project succeeds and stamps identity; mismatched identity rolls back the imported repo); runtime tests (cross-host project setup, rollback on failed existing-folder and clone alignment).

## AI Review Report

I reviewed this diff for correctness and edge cases.

- **Guard replacement is sound.** The old guard blocked any project not yet in the local store. The new check `getProjectIdForProviderIdentity(identity) !== projectId` is stricter in a good way: it verifies the supplied identity actually maps to the requested project id via the same normalization used for projection, so a spoofed/mismatched identity still fails with `Imported folder does not match the selected project identity.` The `repos-remote` rollback test confirms a mismatch (`owner: 'other'`) is rejected and the repo removed.
- **Atomicity / rollback.** Both the local handler and the runtime service only roll back when the repo was newly created (`!result.alreadyExisted` / `repoWasCreated`), correctly avoiding removal of a pre-existing repo. The runtime path additionally invalidates resolved-worktree and worktree-scan caches and notifies renderer — consistent with other mutation paths. The `repoWasCreated` flag is computed from a pre-import snapshot of known repo ids, which is the right signal.
- **Clone path refactor.** `setupProjectClone` previously delegated to `setupProjectExistingFolder`; it now shares `completeProjectHostSetup`, so clone imports gain the same rollback guarantee (covered by the new clone-rollback test). Behavior for the success path is preserved (`setupMethod: 'cloned'`, `path: repo.path`).
- **Identity scope.** The threaded `projectProviderIdentity` is constrained to `provider: 'github'` in the zod schemas and types; non-GitHub (`git:` / `repo:`) ids that don't match the imported folder still fail cleanly. No widening of accepted input beyond GitHub identities.
- **Cross-platform (macOS/Linux/Windows).** I explicitly checked for platform-sensitive concerns. No keyboard shortcuts, shortcut labels, path separators, shell invocation, or Electron-specific APIs are added or changed. Path handling is unchanged (paths flow through existing repo import). Tests use `join`/`tmpdir` for OS-neutral temp paths. Nothing platform-specific was touched.

No fabricated findings; the above is based solely on the diff.

## Security Audit

- **Input handling / IPC.** A new optional `projectProviderIdentity` field is added to the `projectHostSetups:setupExistingFolder` IPC and the runtime RPC schemas. It is validated with zod (`provider` fixed to the `github` literal; non-empty `owner`/`repo`; optional non-empty `host`), so untrusted callers cannot inject arbitrary shapes.
- **Authorization / trust.** The identity is not blindly trusted: `getProjectIdForProviderIdentity(identity)` must equal the requested `projectId`, so a client cannot use a forged identity to link a repo to an unrelated project.
- **Filesystem / roots.** Authorized-roots cache is invalidated on the rollback path, so a rejected import does not leave a stale authorized root behind — this narrows, not widens, filesystem exposure.
- **Command execution / secrets / dependencies.** No shell/command execution, no auth token or secret handling, and no dependency changes are introduced.

No new security risks identified in this diff.

## Notes

- SSH use case: the rollback path applies to both local and SSH/runtime imports; identity is carried across the local↔runtime boundary rather than assumed local.
- Follow-up: `pnpm build` was not run/claimed by the author; worth confirming in CI before merge.

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

## ELI5

Adding a local machine to a project that only exists on a remote Orca runtime failed with "Project not found." The local setup path now understands runtime-only projects so you can attach a local host successfully.

### CODE REVIEW

File: src/main/runtime/orca-runtime.ts [hunk 1]
```diff
@@ -12594,17 +12595,21 @@ export class OrcaRuntimeService {
...
+      // store), so fall back to the identity encoded in the project id when this
+      // runtime has no project record to read `providerIdentity` from.
+      const existingProject = this.listProjects().find(
+        (project) => project.id === args.projectId
+      )
+      const identity =
+        existingProject?.providerIdentity ?? parseProjectProviderIdentity(args.projectId)
+      if (!identity || identity.provider !== 'github') {
         throw new Error('Imported folder does not match the selected project identity.')
       }
```
[the remark this answers is not in the document]
Comment (PannenetsF) [reply src/main/runtime/orca-runtime.ts]: Good catch — addressed in 9c147e382. `setupProjectExistingFolder` now snapshots the repo ids before `addRepo`, and if alignment/linking fails it removes the freshly-created repo (and re-notifies), matching the atomic rollback in the local IPC handler. An already-existing repo is never removed. Added a regression test that imports a folder for a non-GitHub project id that can't be reconstructed and asserts the runtime store ends up empty.

**Comment by brennanb2025** [comment]
Taking a look!

File: src/main/runtime/orca-runtime.ts [hunk 2]
```diff
@@ -12591,42 +12592,59 @@ export class OrcaRuntimeService {
...
+        throw new Error(`Project setup was created without a project record: ${setup.projectId}`)
+      }
+      return { project, setup, repo }
+    } catch (err) {
+      if (repoWasCreated) {
+        this.store.removeProject?.(repo.id)
+        this.invalidateResolvedWorktreeCache()
+        this.notifyReposChanged()
+      }
+      throw err
```
[the remark this answers is not in the document]
Comment (PannenetsF) [reply src/main/runtime/orca-runtime.ts:15396]: Addressed in 612110aaf — the rollback now mirrors the canonical `removeProject` cleanup: it also calls `invalidateWorktreeScanCacheForRepo(repo.id)` and `invalidateAuthorizedRootsCache()` after `store.removeProject`, so no stale worktree-scan or authorized-roots cache entry survives the rolled-back import.

**Comment by brennanb2025** [comment]
Updated this branch and rebased it onto current `main`. I replaced the project-ID reverse parsing with an explicit, validated provider-identity payload carried from the source host, so runtime-only projects can be added locally without losing host-qualified identity or accepting mismatched IDs. I also unified atomic rollback for local/runtime imports and clone setup, including the canonical cache invalidations requested in review. Validation is clean: 1,316 affected tests, node/web/CLI typechecks, touched-file lint, and the max-lines ratchet all pass.

**Review by brennanb2025 — APPROVED** [review]
Thanks for this! Nice catch on the asymmetry between the local IPC handler and the runtime handler. Carrying the structured provider identity across the host boundary (rather than parsing it back out of the project id) is the right call, and making the import atomic so a mismatched folder rolls back instead of orphaning a repo is a good hardening.

Verified locally: the affected suites (`project-host-setup-projection`, `repos-remote`, `orca-runtime`, renderer repos slice) pass, plus `tsc` (node/cli/web) and `oxlint` clean. The handler-level integration tests exercise both the runtime-only-project success path and the mismatch rollback path directly. LGTM.

**Review by brennanb2025 — APPROVED** [review]
Thanks for this! Nice catch on the asymmetry between the local IPC handler and the runtime handler. Carrying the structured provider identity across the host boundary (rather than parsing it back out of the project id) is the right call, and making the import atomic so a mismatched folder rolls back instead of orphaning a repo is a good hardening.

Verified locally: the affected suites (`project-host-setup-projection`, `repos-remote`, `orca-runtime`, renderer repos slice) pass, plus `tsc` (node/cli/web) and `oxlint` clean. Also reproduced the original scenario end-to-end against a live paired `orca serve` runtime — a runtime-only `github:pytorch/pytorch` project with no local record — and adding a Local host now succeeds with the identity stamped correctly, where it previously failed with "Project not found". LGTM.
